build: Add unit test build control through `BUILD_TESTING` and `YSTDLIB_CPP_BUILD_TESTING` flags; Support CTest through Catch2 functions.

Bill-hbrhbr opened the pull request:
<!-- markdownlint-disable MD012 -->

<!--
Set the PR title to a meaningful commit message that:

* is in imperative form.
* follows the Conventional Commits specification (https://www.conventionalcommits.org).
  * See https://github.com/commitizen/conventional-commit-types/blob/master/index.json for possible
    types.

Example:

fix: Don't add implicit wildcards ('*') at the beginning and the end of a query (fixes #390).
-->

# Description

<!-- Describe what this request will change/fix and provide any details necessary for reviewers. -->

When including `ystdlib-cpp` as a subproject, all options should be prefixed with `YSTDLIB_CPP_` to avoid affecting parent project settings.

Regarding unit test building, we want to respect the `BUILD_TESTING` flag of the parent project. However, if users wish to turn off unit test building only for `ystdlib-cpp`, they can specify
```
option(YSTDLIB_CPP_BUILD_TESTING "" OFF)
```
before including `ystdlib-cpp` as a subproject.

We also bring in `CTest` framework since it's easy to integrate it with `Catch2`'s `catch_discover_tests()`.

# Checklist

<!-- Ensure each item below is satisfied and indicate so by inserting an `x` within each `[ ]`. -->

* [X] The PR satisfies the [contribution guidelines][yscope-contrib-guidelines].
* [X] This is a breaking change and that has been indicated in the PR title, OR this isn't a
  breaking change.
* [X] Necessary docs have been updated, OR no docs need to be updated.

# Validation performed

<!-- Describe what tests and validation you performed on the change. -->

- [X] No change to functionalities.

[yscope-contrib-guidelines]: https://docs.yscope.com/dev-guide/contrib-guides-overview.html


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Summary by CodeRabbit

- **Tests**
	- Revised the unit testing configuration for enhanced flexibility when building and executing tests.
- **Documentation**
	- Updated integration instructions to include a new section on using CMake and suggested disabling the building of unit tests.
- **Chores**
	- Refined version control settings to automatically ignore test-related directories and files.
- **Configuration**
	- Updated minimum required CMake version and introduced new options for controlling test builds.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

kirkrodrigues commented on CMakeLists.txt:
```diff
@@ -11,12 +11,12 @@ set(CMAKE_EXPORT_COMPILE_COMMANDS
 )
 
-option(BUILD_TESTING "If ON, unit tests will be built." ON)
+option(YSTDLIB_CPP_BUILD_TESTING "If ON, unit tests will be built." ON)
```
Isn't [BUILD_TESTING](https://cmake.org/cmake/help/git-stage/variable/BUILD_TESTING.html) a common CMake option? Can read [here](https://cliutils.gitlab.io/modern-cmake/chapters/testing.html) for a bit more context.

Sidenote: Any reason we're not using [CTest](https://cmake.org/cmake/help/latest/manual/ctest.1.html)?

Bill-hbrhbr replied:
let me read thru these links

Bill-hbrhbr replied:
Upon a quick scan through, there's nothing that we need that can be achieved by `CTest` but cannot by `CMake` + `Task`.
But it's good to read the common practice regarding `BUILD_TESTING`, which I'm gonna reference while updating this PR.

kirkrodrigues commented:
@coderabbitai review

davidlion reviewed (changes requested):
1. integrate with ctest since it is so easy https://github.com/catchorg/Catch2/blob/devel/docs/cmake-integration.md#usage
2. simplify the flag logic a bit: follow `BUILD_TESTING` but allow an override through `YSTDLIB_CPP_BUILD_TESTING`
    1. default `YSTDLIB_CPP_BUILD_TESTING` to `ON`
    2. `if (NOT (YSTDLIB_CPP_BUILD_TESTING AND BUILD_TESTING)) set(YSTDLIB_CPP_ENABLE_TESTS OFF)` (if both flags are set run unit tests if either is off we turn off enable_tests)
        1. as a side-effect we don't need to set/modify `BUILD_TESTING` (and can leave it to the user setting)
        2. can drop `YSTDLIB_CPP_MASTER_PROJECT`
        3. use `YSTDLIB_CPP_ENABLE_TESTS` same as before

davidlion commented on README.md:
Lets add a small section to `##Testing` that explains the behaviour of `BUILD_TESTING` and `YSTDLIB_CPP_BUILD_TESTING` (currently we never mention that we respect `BUILD_TESTING`). ~2 sentences is probably enough, no need to go into a lot of detail.

Bill-hbrhbr replied:
ok

Bill-hbrhbr commented on CMakeLists.txt:
```diff
@@ -11,25 +11,32 @@ set(CMAKE_EXPORT_COMPILE_COMMANDS
+    set(YSTDLIB_CPP_ENABLE_TESTS ON)
+endif()
+
+# Import CMake helper functions
+list(APPEND CMAKE_MODULE_PATH ${CMAKE_CURRENT_LIST_DIR}/CMake)
```
This is more due to the fact of merging commits

Bill-hbrhbr commented on CMakeLists.txt:
```diff
@@ -11,25 +11,32 @@ set(CMAKE_EXPORT_COMPILE_COMMANDS
-list(APPEND CMAKE_MODULE_PATH ${CMAKE_CURRENT_LIST_DIR}/CMake)
-
-if(PROJECT_SOURCE_DIR STREQUAL CMAKE_SOURCE_DIR)
+if(YSTDLIB_CPP_IS_TOP_LEVEL)
```
Since it's not hard to install a new cmake I will bump it to the default version of `3.22.1` of `ubuntu-jammy`

Bill-hbrhbr commented:
@coderabbitai review

Bill-hbrhbr commented:
@coderabbitai full review

Bill-hbrhbr commented:
> **Actionable comments posted: 0**
> 
> 🧹 Nitpick comments (2)
> > README.md (1)> `7-7`: **Nitpick: Consider Consistent Heading Style**
> > > The heading “## Via CMake's add_subdirectory()” uses the atx style. Static analysis suggests using setext style for consistency. You might consider changing it to a setext-style heading if that aligns with the project’s markdown conventions.
> > > 
> > > 🧰 Tools
> > > 
> > > 🪛 markdownlint-cli2 (0.17.2)
> > > 7-7: Heading style
> > > Expected: setext; Actual: atx
> > > (MD003, heading-style)
> > 
> > 
> > 
> > CMakeLists.txt (1)> `24-27`: **Verify: Conditional Enabling of Unit Tests**
> > > The new option `YSTDLIB_CPP_BUILD_TESTING` is combined with `BUILD_TESTING` to set `YSTDLIB_CPP_ENABLE_TESTS`. Please verify that this logic meets the intended behaviour when `BUILD_TESTING` is either undefined or set to ON in parent projects. A brief clarification in the accompanying documentation might help users understand the condition when ystdlib-cpp is added as a subproject.
> 
> 📜 Review details
> **Configuration used: CodeRabbit UI** **Review profile: CHILL** **Plan: Pro**
> 
> 📥 Commits
> 📒 Files selected for processing (2)
> * `CMakeLists.txt` (3 hunks)
> * `README.md` (2 hunks)
> 
> 🧰 Additional context used
> ⏰ Context from checks skipped due to timeout of 90000ms (1)
> 🔇 Additional comments (6)
> > README.md (1)
> > `67-70`: **Approved: Clear Explanation of Unit Test Configuration**
> > The new text clearly explains that as a top-level CMake project, `ystdlib-cpp` always builds unit tests, and when included as a subproject it respects the parent-scope `BUILD_TESTING` flag. This enhancement addresses the earlier review comment about documenting the interplay between `BUILD_TESTING` and `YSTDLIB_CPP_BUILD_TESTING`. If feasible, you might also consider restating this explanation in the “## Testing” section to further assist users.
> > 
> > 
> > CMakeLists.txt (5)
> > `1-1`: **Approved: Updated Minimum Required CMake Version**
> > The minimum required CMake version has been bumped to 3.22.1, which is appropriate given the newer features and improved handling (e.g. using `PROJECT-NAME_IS_TOP_LEVEL`). Ensure this change is clearly documented for users upgrading from older versions.
> > 
> > `14-22`: **Approved: Conditional Inclusion of CTest**
> > Using `if(YSTDLIB_CPP_IS_TOP_LEVEL)` to conditionally include the CTest module is a sensible design choice. This ensures that dependency settings and testing modules are only activated when the project is the top-level project.
> > 
> > `32-40`: **Approved: Catch2 Integration for Unit Testing**
> > The block that checks `if(YSTDLIB_CPP_ENABLE_TESTS)` and then finds Catch2 (version 3.8.0) is well implemented. The informative status message and fatal error in case of failure provide clear feedback.
> > 
> > `41-54`: **Approved: Unified Unit Test Target Setup**
> > The creation of the unified unit test target—including linking with `Catch2::Catch2WithMain`, setting the C++ standard, and configuring the output directory—is well implemented. The use of `catch_discover_tests` further facilitates automated test discovery.
> > 
> > `56-59`: **Approved: Inclusion of Helper Functions and Subdirectory**
> > The final inclusion of `ystdlib-cpp-helpers` and adding the subdirectory for `src/ystdlib` integrate the library’s structure coherently.

I already added doc to explain build testing flags.

davidlion commented on README.md:
````diff
@@ -60,6 +63,17 @@ To build and run unit tests for a specific library:
 task test-<lib_name>
 ```
 
+The default behaviors for ystdlib-cpp unit test building are:
+* As a top-level project, unit tests are built.
+* When included as a subproject, the parent-scope `BUILD_TESTING` flag (true /false/undefined) is
+  strictly respected.
+If you do not wish to build unit tests under any circumstances, you can override the default
+behaviors by setting CMake option `YSTDLIB_CPP_BUILD_TESTING` to `OFF`.
+
+As a top-level CMake project, `ystdlib-cpp` always builds unit tests. When included as a subproject,
+it respects the parent-scope `BUILD_TESTING` flag. In either case, you can turn off unit test
+building by setting option `YSTDLIB_CPP_BUILD_TESTING` to off.
````
You forgot to remove the old text and there is an extra space in the 2nd bullet. I also realized we probably shouldn't say `OFF` directly since cmake supports a handful of true/false values.

Here is my suggestion to simplify the text:
```suggestion
When generating testing target the CMake variable `BUILD_TESTING` is followed, unless overruled by
setting `YSTDLIB_CPP_BUILD_TESTING` to false. By default, if built as a top-level project
`BUILD_TESTING` will be set to true and unit tests will be built.
```

Bill-hbrhbr commented:
@coderabbitai I thought the README already uses ATX style markdown

Bill-hbrhbr commented:
@coderabbitai You no longer comment directly on the files, so how can I 
`For best results, initiate chat on the files or code changes.`
